sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala

```diff
@@ -544,6 +597,42 @@ private[sql] class HDFSBackedStateStoreProvider extends StateStoreProvider with
     result
   }
 
+  private def loadMap(startVersion: Long, endVersion: Long): HDFSBackedStateStoreMap = {
+
+    val (result, elapsedMs) = Utils.timeTakenMs {
+      val startVersionMap = Option(loadedMaps.get(startVersion)) match {
+        case Some(value) =>
+          loadedMapCacheHitCount.increment()
+          Option(value)
+        case None =>
+          logWarning(
+            log"The state for version ${MDC(LogKeys.FILE_VERSION, startVersion)} doesn't " +
+            log"exist in loadedMaps. Reading snapshot file and delta files if needed..." +
+            log"Note that this is normal for the first batch of starting query.")
+          loadedMapCacheMissCount.increment()
+          readSnapshotFile(startVersion)
+      }
+      if (startVersionMap.isEmpty) {
+        throw StateStoreErrors.stateStoreSnapshotFileNotFound(
+          snapshotFile(startVersion).toString, toString())
+      }
+      synchronized { putStateIntoStateCacheMap(startVersion, startVersionMap.get) }
```

Comment:
Why do we need to do this at all ? This is only acting as a cache right that really only is useful in the context of streaming queries ?

Reply:
For the purpose of state reader, we do not need it. However, I was thinking to make this method general (like the other loadMap). I kept this caching action to be consistent with the other loadMap.

Reply:
Yea - as mentioned in the comment below. By default I believe we keep only 2 versions in this map. given that the start version might be older - its likely that we just add it to the map and remove it soon or skip adding to the map at all. also, its unlikely that we will reference this version in the future. so we can keep it for code consistency if needed, but we should see whether its possible for it to cause side-effects for a running query using the provider. also - lets explain the path we are taking as a code comment